src/main/java/com/minecolonies/core/placementhandlers/SolidPlaceholderPlacementHandler.java

```diff
@@ -0,0 +1,119 @@
+import net.minecraft.world.level.block.Blocks;
+import net.minecraft.world.level.block.state.BlockState;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.ldtteam.structurize.api.util.constant.Constants.UPDATE_FLAG;
+
+public class SolidPlaceholderPlacementHandler implements IPlacementHandler
+{
+    /**
+     * Replacement block for solid placeholders
+     */
+    private BlockState replacement = Blocks.DIRT.defaultBlockState();
+
+    private IPlacementHandler replacementHandler = null;
+
+    /**
+     * Sets a different replacement block
+     *
+     * @param state
+     */
+    public void setReplacement(final BlockState state)
+    {
+        replacement = state;
+    }
+
+    /**
+     * Get the replacement block
+     *
+     * @return
+     */
+    public BlockState getReplacement()
+    {
+        return replacement;
+    }
+
+    @Override
+    public boolean canHandle(Level world, BlockPos pos, BlockState blockState)
+    {
+        return blockState.getBlock() instanceof BlockSolidSubstitution;
+    }
+
+    private void searchHandler(final Level world, final BlockPos pos)
+    {
+        if (replacementHandler == null)
+        {
+            for (final IPlacementHandler handler : PlacementHandlers.handlers)
+            {
+                if (handler.canHandle(world, pos, replacement))
+                {
+                    replacementHandler = handler;
+                    break;
+                }
+            }
+        }
+    }
+
+    @Override
+    public List<ItemStack> getRequiredItems(
+        Level world,
+        BlockPos pos,
+        BlockState blockState,
+        @Nullable CompoundTag tileEntityData,
+        boolean complete)
+    {
+        searchHandler(world, pos);
+        List<ItemStack> items = new ArrayList<>();
+
+        if (complete)
+        {
+            // for scan tool, show the actual placeholder block
+            items.add(new ItemStack(blockState.getBlock()));
+        }
+        else
+        {
+            return replacementHandler.getRequiredItems(world, pos, replacement, tileEntityData, complete);
+        }
+
+        return items;
+    }
+
+    @Override
+    public ActionProcessingResult handle(
+        Level world,
+        BlockPos pos,
+        BlockState blockState,
+        @Nullable CompoundTag tileEntityData,
+        boolean complete,
+        BlockPos centerPos, final PlacementSettings settings)
+    {
+        if (complete)
+        {
+            world.setBlock(pos, ModBlocks.blockSubstitution.get().defaultBlockState(), UPDATE_FLAG);
+            return ActionProcessingResult.PASS;
+        }
+
+        if (BlockUtils.isAnySolid(world.getBlockState(pos)))
+        {
+            return ActionProcessingResult.DENY;
+        }
+
+        searchHandler(world, pos);
+        return replacementHandler.handle(world, pos, replacement, tileEntityData, complete, centerPos, settings);
```

Comment:
maybe some sort of loop protection?

Reply:
could add it, though atm its not needed the placement handler is not added to the PlacementHandlers.handlers, it is just added to a local list during block search

Reply:
Ah okay, it's not global

Reply:
added it anyways, since I guess we eventually should use it as proper handling of placeholders

Reply:
<3